### PR TITLE
Skeleton parameters sync API

### PR DESCRIPTION
**Pull request by esteve**
Connects to ros2/ros2#28

@dirk-thomas @tfoote @wjwwood

### CODE REVIEW

File: rclcpp/include/rclcpp/parameter.hpp [hunk 1]
```diff
@@ -56,31 +57,43 @@ class ParameterVariant
...
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_DOUBLE;
+  }
+  explicit ParameterVariant(const std::string & name, const std::string & string_value)
   : name_(name)
   {
     value_.string_value = string_value;
     value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_STRING;
   }
-  ParameterVariant(const std::string & name, const std::vector<int8_t> & bytes_value)
+  explicit ParameterVariant(const std::string & name, const std::vector<int8_t> & bytes_value)
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: Should the bytes be signed or unsigned? I've done unsigned in the past, but I'm not sure that it matters.

**Comment by tfoote** [comment]
Overall this looks good. I'm a little worried about implicitly creating the executor inside. This is something we should review going forward. But we can use this for now. 

+1

File: rclcpp/include/rclcpp/parameter.hpp (hunk 1 above)
Comment (esteve) [inline rclcpp/include/rclcpp/parameter.hpp]: This matches the code generated in `ParameterValue_Struct.h` from `ParameterValue.msg`, but I'd prefer `uint8_t` myself.

Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: It should indeed be `uint8` to follow our message parser which defines all the types: https://github.com/ros2/rosidl/blob/master/rosidl_parser/rosidl_parser/__init__.py#L484-L485

Probably the `ParameterValue` message needs to be updated to change `bytes_value` to `int8[]`.

File: rclcpp/include/rclcpp/parameter.hpp [hunk 2]
```diff
@@ -233,6 +248,72 @@ class AsyncParametersClient
   rclcpp::client::Client<rcl_interfaces::DescribeParameters>::SharedPtr describe_parameters_client_;
 };
 
+class SyncParametersClient
+{
+  friend class rclcpp::executor::Executor;
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Why is `friend` needed here? It seems to only use the public constructor.

Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter.hpp]: Executor may need private/protected members in this class? Not sure though.

File: rclcpp/include/rclcpp/parameter.hpp (hunk 1 above)
Comment (esteve) [inline rclcpp/include/rclcpp/parameter.hpp]: I replaced `int8[]` with `byte[]` in `ParameterValue` as it's semantically more accurate:https://github.com/ros2/rcl_interfaces/commit/ee983f35267f7d27d945b3cd9d6fd4bb57243eaa

File: rclcpp/include/rclcpp/parameter.hpp (hunk 2 above)
Comment (esteve) [inline rclcpp/include/rclcpp/parameter.hpp]: Removed. This was a remnant of a quick test.

**Comment by esteve** [comment]
I've renamed ParameterListResult and ParameterSetResult to ListParametersResult and SetParametersResult respectively in ros2/rcl_interfaces#3

**Comment by dirk-thomas** [comment]
Currently the API uses multiple template specializations of `get_value` with different return types. So the user must always specify the template type. Do we want to provide methods which contain the type in the name (additionally or instead)?

Also I think we could still return `const std::string &` in the case of strings (https://github.com/ros2/rclcpp/pull/26#discussion_r29351986). I don't think that the `template T` prevents us from specializing it for `const std::string &`?

**Comment by esteve** [comment]
@dirk-thomas I've changed it so that the user only needs to pass the `ParameterType` value to `get_value<>` instead of using types.

I think this could be improved if we didn't require the storage to be a `ParameterValue`, which is completely an implementation detail. Then we could declare all the constructors as `constexpr` as well as `get_type` and it'll be less cumbersome:

`auto value = p.get_value<p.get_type()>()`

or we might even get rid of `p.get_type` and just use `p.get_value`

**Comment by wjwwood** [comment]
There's precedence in other API's for have something like `.asString`, `.asInt`, and `.asDouble` rather than `.as<std::string>`, `.as<int64_t>`, and `.as<double>`, especially when there are a finite set of options. I don't really mind one way or the other.

As for it being aligned with the style of the C api, I think it's fine if they are not one to one. If it makes more sense to use a template in the C++ and an expanded name in C that's fine. It can be made similar and familiar without being exactly the same.

**Comment by esteve** [comment]
+1 to `as_string`, `as_int`, etc. instead of templatized `get_value`

**Comment by esteve** [comment]
I've added the `as_*` versions of the `get_value` method, so when we reach a consensus I'll just remove the one that less popular. My vote goes for the `as_*` methods.

File: rclcpp/include/rclcpp/parameter.hpp [hunk 3]
```diff
@@ -101,61 +109,78 @@ class ParameterVariant
...
+      throw std::runtime_error("Invalid type");
+    }
+    return value_.bytes_value;
+  }
+
+  int64_t as_int() const {return get_value<ParameterType::PARAMETER_INTEGER>(); }
+
+  double as_double() const {return get_value<ParameterType::PARAMETER_DOUBLE>(); }
+
+  const std::string & as_string() const {return get_value<ParameterType::PARAMETER_STRING>(); }
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: With `get_value<>` not returning a const reference but a temporary object this won't work I guess?

Comment (esteve) [inline rclcpp/include/rclcpp/parameter.hpp]: `get_value<>` returns a const reference https://github.com/ros2/rclcpp/pull/28/files#diff-b54f433d1e964fc09df27a3262119169R133

Comment (dirk-thomas) [inline rclcpp/include/rclcpp/parameter.hpp]: Should we do the same for `std::vector<uint8_t>`?

Comment (esteve) [inline rclcpp/include/rclcpp/parameter.hpp]: Done.

**Pull request merged**